System2 pendulum dynamic constraint

sammy-tri opened the pull request:
Converting Pendulum is still a WIP at this point (for example, pendulum_run_dynamics and pendulum_dynamic_constraint_test have been converted to System2, but no other tests have yet).

Filing this PR early in part to get it reviewed for idiomatic System2 usage before bashing along through the other examples (since it's my first System2 work...)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3631)
<!-- Reviewable:end -->

sammy-tri commented:
+@naveenoid for feature review, +@jwnimmer-tri for platform review (and because he kinda volunteered to take a look at this in a f2f discussion a couple of days ago)

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KSrdvizEuXxbF-B9Cnp:b-gjug3a)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Checkpoint (build system reviewed).

Also, please rebase to latest master, so I can confirm the CMakeList edits versus the new style unit test declarations.

---

Reviewed 3 of 9 files at r1.
Review status: 3 of 9 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/examples/Pendulum/CMakeLists.txt, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSv8q0Ds6mVRqc3dyur:-KSv8q0Ds6mVRqc3dyus:bvwf95g) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/CMakeLists.txt#L22)):*
> ```CMake
>     LIBS -ldrakePendulumSystem
>     REQUIRES
>     VERSION 0.0.1)
> ```

FYI Not this PR's fault, but declaring PODS and pkg_config glue for a example-only single-file static (oops, dynamic!) library makes me weep.

---

*[drake/examples/Pendulum/CMakeLists.txt, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSv8H_WTrThJKB2Ol9v:-KSv8H_WTrThJKB2Ol9w:b-l8v9yr) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/CMakeLists.txt#L28)):*
> ```CMake
>     drakePendulumSystem
>     drakeRigidBodyPlant
>     drakeRBM
> ```

BTW Redundant

---

*[drake/examples/Pendulum/CMakeLists.txt, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSv8GEwu5g3-Kl67Toh:-KSv8GExQbeQoXfc2NtF:b-l8v9yr) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/CMakeLists.txt#L29)):*
> ```CMake
>     drakeRigidBodyPlant
>     drakeRBM
>     drakeSystemAnalysis
> ```

BTW Redundant

---

*[drake/examples/Pendulum/CMakeLists.txt, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSv8JbmncGpDE5iXLcj:-KSv8JbmncGpDE5iXLck:b-l8v9yr) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/CMakeLists.txt#L30)):*
> ```CMake
>     drakeRBM
>     drakeSystemAnalysis
>     threads)
> ```

BTW Redundant

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KSv8YRQ4BqZngwTI7G8:bjkdoec)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Checkpoint (constraints).

---

Reviewed 3 of 9 files at r1.
Review status: 6 of 9 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[drake/examples/Pendulum/test/pendulum_dynamic_constraint_test.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvESGliAe_L21uz99M:-KSvESGliAe_L21uz99N:b-xli7uh) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/test/pendulum_dynamic_constraint_test.cc#L27)):*
> ```cpp
> 
>   drake::systems::System2DirectCollocationConstraint dut(
>       std::make_unique<PendulumSystem<drake::AutoDiffXd>>());
> ```

FYI super-cool that the test just ports forward easily like this.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.cc, line 90 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvCdQH4o5nZK78pqLQ:-KSvCdQH4o5nZK78pqLR:b-jms2kc) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/systems/plants/constraint/direct_collocation_constraint.cc#L90)):*
> ```cpp
>           system_->get_input_port(0).get_size()));
>   std::unique_ptr<InputPort> input_port(input_port_);
>   context_->SetInputPort(0, std::move(input_port));
> ```

FYI This is microscopically cleaner IMO:

```
auto input_port = std::make_unique<FreestandingInputPort>(
    std::make_unique<BasicVector<AutoDiffXd>>(
        system_->get_input_port(0).get_size()));
input_port_ = input_port.get();
context_->SetInputPort(0, std::move(input_port));
```

Perhaps I can convince System2 authors to offer a templated SetInputPort that would allow doing this in the initalizer list also (if SetInputPort returned the port alias).

---

*[drake/systems/plants/constraint/direct_collocation_constraint.cc, line 103 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvC920oymzIZGXhZ_Q:-KSvC9218qnyMbKrGLk2:blpkebk) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/systems/plants/constraint/direct_collocation_constraint.cc#L103)):*
> ```cpp
>       ->get_mutable_state()->SetFromVector(state);
>   std::unique_ptr<ContinuousState<AutoDiffXd>> derivatives =
>       system_->AllocateTimeDerivatives();
> ```

BTW Unclear why `input_port_` is reused as a member field but `derivatives` is not.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.cc, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvBCy5kx2XaMWxVUOJ:-KSvBCy6bhIeXqke4raE:b4w2rmf) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/systems/plants/constraint/direct_collocation_constraint.cc#L109)):*
> ```cpp
> 
> ```

BTW extra vertical whitespace

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 86 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvAGWzgv7TLuFOC-fE:-KSvAGWzgv7TLuFOC-fF:b1sfg8y) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/systems/plants/constraint/direct_collocation_constraint.h#L86)):*
> ```objc
>     : public DirectCollocationConstraint {
>  public:
>   explicit System2DirectCollocationConstraint(
> ```

Document the argument?  Is any system whose type matches sensible to pass in?

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSv9ppjjSVKSaN0V9Yu:-KSv9ppkRrP5V9bVayuI:b-q2m3an) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/systems/plants/constraint/direct_collocation_constraint.h#L88)):*
> ```objc
>   explicit System2DirectCollocationConstraint(
>       std::unique_ptr<System<AutoDiffXd>>);
>   virtual ~System2DirectCollocationConstraint();
> ```

override

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 97 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvAjpnBvjFawybGZi4:-KSvAjpnBvjFawybGZi5:b-c1twdz) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/systems/plants/constraint/direct_collocation_constraint.h#L97)):*
> ```objc
>   std::unique_ptr<System<AutoDiffXd>> system_;
>   std::unique_ptr<Context<AutoDiffXd>> context_;
>   FreestandingInputPort* input_port_;
> ```

POD member field, so value-init required at point of declaration.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KSvEXq5mBIt7wNHSUrK:bukae9y)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
First pass complete.

---

Reviewed 3 of 9 files at r1.
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[drake/examples/Pendulum/pendulum_run_dynamics.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvEwzGCiC_QmK44HOK:-KSvEwzGCiC_QmK44HOL:b-xvaixb) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_run_dynamics.cc#L12)):*
> ```cpp
> #include "drake/systems/plants/RigidBodyTree.h"
> 
> using namespace std;
> ```

Disallowed per cppguide.

---

*[drake/examples/Pendulum/pendulum_run_dynamics.cc, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvEz6f495pXrGqgglG:-KSvEz6f495pXrGqgglH:b-xvaixb) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_run_dynamics.cc#L13)):*
> ```cpp
> 
> using namespace std;
> using namespace drake;
> ```

Disallowed per cppguide.

---

*[drake/examples/Pendulum/pendulum_run_dynamics.cc, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvFKDu4YDhSFk-Gp_b:-KSvFKDu4YDhSFk-Gp_c:b-ddm761) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_run_dynamics.cc#L29)):*
> ```cpp
>       builder.AddSystem<systems::RigidBodyTreeLcmPublisher>(tree, &lcm);
>   builder.Connect(source->get_output_port(), pendulum->get_tau_port());
>   builder.Connect(pendulum->get_output_port(), publisher->get_input_port(0));
> ```

BTW Would you have preferred to have `publisher->get_input_port()` available to you?  I could put that on the feature queue for that System.

---

*[drake/examples/Pendulum/pendulum_run_dynamics.cc, line 35 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvFf2_snpqeZi148Mc:-KSvFf2_snpqeZi148Md:b-dw5zu2) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_run_dynamics.cc#L35)):*
> ```cpp
>   pendulum->set_theta(
>       diagram->GetMutableSubsystemContext(
>           simulator.get_mutable_context(), pendulum), 1.);
> ```

@david-germani-tri Is this the correct way to establish the initial state of diagram?  (Frob the context, then `Initialize()`.)

---

*[drake/examples/Pendulum/pendulum_system.cc, line 91 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvRZ3jD9b0SW5Z8UBk:-KSvRZ3k6neQW3Pm4DSs:b-ddbjr1) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_system.cc#L91)):*
> ```cpp
>   }
>   return std::unique_ptr<systems::SystemOutput<T>>(output.release());
> }
> ```

I believe that overriding `AllocateOutputVector` is more maintainable (less duplicated code), and still meets the needs of this System.  Try it?

---

*[drake/examples/Pendulum/pendulum_system.cc, line 98 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvRllBKkrSnFKtuhYi:-KSvRllBKkrSnFKtuhYj:b4rv3r7) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_system.cc#L98)):*
> ```cpp
>   return std::make_unique<systems::ContinuousState<T>>(
>       std::make_unique<PendulumStateVector<T>>(),
>       1 /* num_q */, 1 /* num_v */, 0 /* num_z */);
> ```

BTW Consider `static_assert(kStateSize == 1 + 1);` or similar?

---

*[drake/examples/Pendulum/pendulum_system.cc, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvSYFcYIxi29oW3sw9:-KSvSYFcYIxi29oW3swA:bb7teo4) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_system.cc#L108)):*
> ```cpp
>   PendulumStateVector<T>* output_vector = get_mutable_output(output);
>   output_vector->set_theta(state.get_theta());
>   output_vector->set_thetadot(state.get_thetadot());
> ```

BTW Wouldn't directly setting all of `output_vector` via all of `state` be clearer and more robust?

FYI @david-german-tri A sugar for "my first output port is always my state" is desperately needed; most q,v systems will do that, and we're already well past the rule-of-3 for that pattern.

---

*[drake/examples/Pendulum/pendulum_system.cc, line 118 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvTKTZTWhCXN4-IOLh:-KSvTKTZTWhCXN4-IOLi:b6uh9c3) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_system.cc#L118)):*
> ```cpp
>   DRAKE_ASSERT_VOID(systems::System<T>::CheckValidContext(context));
> 
>   // TODO(david-german-tri): Cache the output of this function.
> ```

BTW Do we really need these here (and above)?

---

*[drake/examples/Pendulum/pendulum_system.cc, line 125 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvTYxv0FW2iHFJsTcU:-KSvTYxv0FW2iHFJsTcV:bsr28jm) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_system.cc#L125)):*
> ```cpp
>   derivative_vector->set_thetadot(
>       (get_tau(context) - m_ * g_ * lc_ * sin(state.get_theta()) -
>        b_ * state.get_thetadot()) / I_);
> ```

BTW Consider a citation to background reading.

---

*[drake/examples/Pendulum/pendulum_system.h, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvP5yNds8fuzaKCKE3:-KSvP5yNds8fuzaKCKE4:b-d0vtey) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_system.h#L32)):*
> ```objc
>   T get_thetadot() const;
>   void set_thetadot(const T& thetadot);
> ```

Disable copy and move, per cppguide.

---

*[drake/examples/Pendulum/pendulum_system.h, line 49 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvQJPtZ0AMtGFv73LL:-KSvQJPuhsgd-IN0OcZQ:b-q2m3an) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_system.h#L49)):*
> ```objc
>  public:
>   PendulumSystem();
>   ~PendulumSystem();
> ```

override

---

*[drake/examples/Pendulum/pendulum_system.h, line 93 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvP76wq22q7HG_4iU0:-KSvP76xyuM7y-kX88i1:b-d0vtey) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_system.h#L93)):*
> ```objc
>   void EvalTimeDerivatives(const MyContext& context,
>                            MyContinuousState* derivatives) const override;
> ```

Disable copy and move, per cppguide.

---

*[drake/examples/Pendulum/pendulum_system.h, line 126 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvTAtuxrpnZgID_ClZ:-KSvTAtuxrpnZgID_Cl_:bsx0dkd) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_system.h#L126)):*
> ```objc
>   static PendulumStateVector<T>* get_mutable_state(MyContext* context) {
>     return get_mutable_state(context->get_mutable_continuous_state());
>   }
> ```

TBD (for myself) I want to revisit all of these (private and public) helpers once other review passes are done.

---

*[drake/examples/Pendulum/pendulum_system.h, line 133 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvQyveGAL7af8oA1BT:-KSvQyveGAL7af8oA1BU:bi0cat8) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/examples/Pendulum/pendulum_system.h#L133)):*
> ```objc
>   const double lc_{.5};   // m
>   const double I_{.25};   // m*l^2; % kg*m^2
>   const double g_{9.81};  // m/s^2
> ```

Missing citation (for all of these) to the some kind of system model of background reading for what they all mean.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 89 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvP3ZMqEAhJVhZLyUV:-KSvP3ZMqEAhJVhZLyUW:b-d0vtey) ([raw file](https://github.com/robotlocomotion/drake/blob/e54daf5cf51b8b899ddf8ad2cf47ce7c87b2e54c/drake/systems/plants/constraint/direct_collocation_constraint.h#L89)):*
> ```objc
>       std::unique_ptr<System<AutoDiffXd>>);
>   virtual ~System2DirectCollocationConstraint();
> ```

Disable copy and move, per cppguide.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KSvTrOfxlekMjnm1q0O:bvb1a6z)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Rebased to 340cf79

---

Review status: all files reviewed at latest revision, 24 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KSvbn8h_K-e0CFFMCdT:b-7tg1uy)*
<!-- Sent from Reviewable.io -->

billhoffman commented on drake/examples/Pendulum/pendulum_system.h:
```diff
@@ -0,0 +1,138 @@
+/// - double
+/// - AutoDiffXd
+template <typename T>
+class DRAKEPENDULUMSYSTEM_EXPORT PendulumStateVector
```
You can not export a template, you can only export instances of specific types.  Try removing this:  DRAKEPENDULUMSYSTEM_EXPORT.

sammy-tri commented:
Review status: 7 of 9 files reviewed at latest revision, 25 unresolved discussions.

---

*[drake/examples/Pendulum/CMakeLists.txt, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSv8H_WTrThJKB2Ol9v:-KSvem4rJVhnP3wYbT2M:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/CMakeLists.txt#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Redundant
</details>
Done.

---

*[drake/examples/Pendulum/CMakeLists.txt, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSv8GEwu5g3-Kl67Toh:-KSvemXnT68X7cFv4TVA:b-raqg9q) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/CMakeLists.txt#L29)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Redundant
</details>
Actually I don't think it is?  Where else would it come from (also it doesn't link if I remove drakeSystemAnalysis)

---

*[drake/examples/Pendulum/CMakeLists.txt, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSv8JbmncGpDE5iXLcj:-KSvevDclYWpAUyqX-H4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/CMakeLists.txt#L30)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Redundant
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 22 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
> Also, please rebase to latest master, so I can confirm the CMakeList edits versus the new style unit test declarations.

Ah, nevermind.  (Done.)  I forgot that I haven't ported `examples` to use the new macro yet.

---

Review status: all files reviewed at latest revision, 22 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KSw9L3Y4ZAiJmAsAj80:bruseeo)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I'll just assign myself here so I'm in the loop of github notifications. +@amcastro-tri 

---

Review status: all files reviewed at latest revision, 22 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KSwD2MwWaJUTopq_Ds-:b-8bcanw)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 28 unresolved discussions.

---

*[drake/examples/Pendulum/pendulum_system.h, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwB7SQYGxucYIm8VAZ:-KSwB7SQYGxucYIm8VA_:b-kwl328) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L66)):*
> ```objc
>   T get_theta(const MyContext& context) const {
>     return get_state(context).get_theta();
>   }
> ```

Dead code.

---

*[drake/examples/Pendulum/pendulum_system.h, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwB8uxgf-nTVlXaQ_2:-KSwB8uyR5-EH2i7ogFM:bm8lj4a) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L74)):*
> ```objc
>   T get_thetadot(const MyContext& context) const {
>     return get_state(context).get_thetadot();
>   }
> ```

Dead code.

---

*[drake/examples/Pendulum/pendulum_system.h, line 78 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwB9ajVr57pgYzceZL:-KSwB9ajVr57pgYzceZM:bm8lj4a) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L78)):*
> ```objc
>   void set_thetadot(MyContext* context, const T& thetadot) const {
>     get_mutable_state(context)->set_thetadot(thetadot);
>   }
> ```

Dead code.

---

*[drake/examples/Pendulum/pendulum_system.h, line 82 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwCeayag2ZfRplKKqT:-KSwCeayag2ZfRplKKqU:b1y1m4y) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L82)):*
> ```objc
>   T get_tau(const MyContext& context) const {
>     return this->EvalVectorInput(context, 0)->GetAtIndex(0);
>   }
> ```

Unclear why public (and thus lacking API comment).

---

*[drake/examples/Pendulum/pendulum_system.h, line 103 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwCzLmwNp5R79QZ-mm:-KSwCzLmwNp5R79QZ-mn:b-zdh189) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L103)):*
> ```objc
>       const MyContinuousState& cstate) {
>     return dynamic_cast<const PendulumStateVector<T>&>(cstate.get_state());
>   }
> ```

BTW This is an inline private static method, only ever called from one other inline private static method.  Seems a bit of a mouthful / overengineered for that purpose?

---

*[drake/examples/Pendulum/pendulum_system.h, line 113 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwDmmDijhYKg2ZUaIS:-KSwDmmDijhYKg2ZUaIT:bm8lj3a) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L113)):*
> ```objc
>     return dynamic_cast<const PendulumStateVector<T>&>(
>         *output.get_vector_data(0));
>   }
> ```

Dead code

---

*[drake/examples/Pendulum/pendulum_system.h, line 126 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwDciHKUApvuyUPvlC:-KSwDciHKUApvuyUPvlD:b-6jwxg3) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L126)):*
> ```objc
>   static PendulumStateVector<T>* get_mutable_state(MyContext* context) {
>     return get_mutable_state(context->get_mutable_continuous_state());
>   }
> ```

BTW consider moving all private static helper code to the cc file (anonymous functions), so the header is cleaner.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 27 unresolved discussions.

---

*[drake/examples/Pendulum/pendulum_run_dynamics.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvEwzGCiC_QmK44HOK:-KSvg7v2ZVLMhrQY5ACW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_run_dynamics.cc#L12)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Disallowed per cppguide.
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_run_dynamics.cc, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvEz6f495pXrGqgglG:-KSvg8L9aLkYnXPIV2cV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_run_dynamics.cc#L13)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Disallowed per cppguide.
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_run_dynamics.cc, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvFKDu4YDhSFk-Gp_b:-KSvgBmcKo1lZy2P-QrN:bkoxfye) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_run_dynamics.cc#L29)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Would you have preferred to have `publisher->get_input_port()` available to you?  I could put that on the feature queue for that System.
</details>
Maybe?  I don't have any strong feelings on the matter.

---

*[drake/examples/Pendulum/pendulum_system.cc, line 91 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvRZ3jD9b0SW5Z8UBk:-KSw7h372QVcn1-xRoSb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.cc#L91)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I believe that overriding `AllocateOutputVector` is more maintainable (less duplicated code), and still meets the needs of this System.  Try it?
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.cc, line 98 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvRllBKkrSnFKtuhYi:-KSw85f797cLNKMTiKz4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.cc#L98)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider `static_assert(kStateSize == 1 + 1);` or similar?
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.cc, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvSYFcYIxi29oW3sw9:-KSw8nC7dTf8zG9V_Af_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.cc#L108)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Wouldn't directly setting all of `output_vector` via all of `state` be clearer and more robust?
> 
> FYI @david-german-tri A sugar for "my first output port is always my state" is desperately needed; most q,v systems will do that, and we're already well past the rule-of-3 for that pattern.
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.cc, line 118 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvTKTZTWhCXN4-IOLh:-KSw8qBpIgVEsx7zOwLU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.cc#L118)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Do we really need these here (and above)?
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.cc, line 125 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvTYxv0FW2iHFJsTcU:-KSwAC36eSwYdimcszBJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.cc#L125)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider a citation to background reading.
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSrds34RBAnsdEBmsbr-r1-21:-KSvgS0Qrle8gh4nj5MV:b-rns5r6) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.h#L21)):*
<details><summary><i>Previously, billhoffman (Bill Hoffman) wrote…</i></summary>
> You can not export a template, you can only export instances of specific types.  Try removing this:  DRAKEPENDULUMSYSTEM_EXPORT.
</details>
I believe the problem was actually an incorrect export macro in the .cc file.  Thanks!  Also removed extra exports.

---

*[drake/examples/Pendulum/pendulum_system.h, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvP5yNds8fuzaKCKE3:-KSwAjD-EXbU2KeTfnDG:b4bxiyw) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.h#L32)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Disable copy and move, per cppguide.
</details>
Is the fact that these are disabled in the parent class not sufficient?  I guess in other places the derived classes disable again, so will do so here.

---

*[drake/examples/Pendulum/pendulum_system.h, line 49 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvQJPtZ0AMtGFv73LL:-KSwBOpHPkgFEelxUb15:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.h#L49)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> override
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.h, line 93 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvP76wq22q7HG_4iU0:-KSwBPbveIGj0Fky3YOH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.h#L93)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Disable copy and move, per cppguide.
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.h, line 133 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvQyveGAL7af8oA1BT:-KSwBfrC2WI8BuTXDe51:bjzylr9) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.h#L133)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing citation (for all of these) to the some kind of system model of background reading for what they all mean.
</details>
Added citation in C++ implementation.  Should I add it again here?

---

*[drake/examples/Pendulum/pendulum_system.h, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwB7SQYGxucYIm8VAZ:-KSwGCdID8BlqXVdKosb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L66)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Dead code.
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.h, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwB8uxgf-nTVlXaQ_2:-KSwGDH9lGSPchpzu08S:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L74)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Dead code.
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.h, line 78 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwB9ajVr57pgYzceZL:-KSwGDyinyrccdkH2CjA:bv3xkz6) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L78)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Dead code.
</details>
Added usage.

---

*[drake/examples/Pendulum/pendulum_system.h, line 82 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwCeayag2ZfRplKKqT:-KSwGQgZDJQptlmEt6fr:b-exlrps) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L82)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Unclear why public (and thus lacking API comment).
</details>
Done.  (made private)

---

*[drake/examples/Pendulum/pendulum_system.h, line 113 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwDmmDijhYKg2ZUaIS:-KSwHunXI3aj85nvBJtI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L113)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Dead code
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.h, line 126 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwDciHKUApvuyUPvlC:-KSwH2Az-r495xWmobgS:byy6kos) ([raw file](https://github.com/robotlocomotion/drake/blob/3d37f94774657d0fdc9dc7c03af4066a6205a228/drake/examples/Pendulum/pendulum_system.h#L126)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW consider moving all private static helper code to the cc file (anonymous functions), so the header is cleaner.
</details>
While this would clean up the header, I think it would make the usage of the helpers more complicated since the <T> of the class type wouldn't be known.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.cc, line 90 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvCdQH4o5nZK78pqLQ:-KSwBs1HyylTYke_SzGv:b-32ep5b) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/systems/plants/constraint/direct_collocation_constraint.cc#L90)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> FYI This is microscopically cleaner IMO:
> 
> ```
> auto input_port = std::make_unique<FreestandingInputPort>(
>     std::make_unique<BasicVector<AutoDiffXd>>(
>         system_->get_input_port(0).get_size()));
> input_port_ = input_port.get();
> context_->SetInputPort(0, std::move(input_port));
> ```
> 
> Perhaps I can convince System2 authors to offer a templated SetInputPort that would allow doing this in the initalizer list also (if SetInputPort returned the port alias).
</details>
It would be kinda nice to do this from the initializer list.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.cc, line 103 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvC920oymzIZGXhZ_Q:-KSwCjLCD6YchV5GHM9W:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/systems/plants/constraint/direct_collocation_constraint.cc#L103)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Unclear why `input_port_` is reused as a member field but `derivatives` is not.
</details>
Done.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.cc, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvBCy5kx2XaMWxVUOJ:-KSwCkXZbeWk_E6k99G0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/systems/plants/constraint/direct_collocation_constraint.cc#L109)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW extra vertical whitespace
</details>
Done.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 86 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvAGWzgv7TLuFOC-fE:-KSwDACcA4bFp_I6s5n3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/systems/plants/constraint/direct_collocation_constraint.h#L86)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Document the argument?  Is any system whose type matches sensible to pass in?
</details>
Done.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSv9ppjjSVKSaN0V9Yu:-KSwDCHPlOWVTlFeFWoI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/systems/plants/constraint/direct_collocation_constraint.h#L88)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> override
</details>
Done.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 89 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvP3ZMqEAhJVhZLyUV:-KSwDp9eBgO8HKoTmp_N:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/systems/plants/constraint/direct_collocation_constraint.h#L89)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Disable copy and move, per cppguide.
</details>
Done.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 97 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvAjpnBvjFawybGZi4:-KSwEQk7KX-PIBe8luV-:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/systems/plants/constraint/direct_collocation_constraint.h#L97)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> POD member field, so value-init required at point of declaration.
</details>
Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/examples/Pendulum/pendulum_run_dynamics.cc, line 35 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvFf2_snpqeZi148Mc:-KSwgQsvjrrmCYeAjg8A:bdgjw0y) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_run_dynamics.cc#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> @david-germani-tri Is this the correct way to establish the initial state of diagram?  (Frob the context, then `Initialize()`.)
</details>
Oops, meant to tag @david-german-tri here, but typo'd it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Checkpoint.  I've ack'd most discussions, I still need to re-read the new edits.

---

Review status: 3 of 9 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KSwh-4aZbIVDeAnsW9n:b68l8ak)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/plants/constraint/direct_collocation_constraint.cc, line 90 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvCdQH4o5nZK78pqLQ:-KSwhzS-_w03sPjbaDlZ:b-8dnxpp) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/systems/plants/constraint/direct_collocation_constraint.cc#L90)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> It would be kinda nice to do this from the initializer list.
</details>
Filed as #3642.  (Marking this discussion as resolved.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 3 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Pendulum/pendulum_system.h, line 133 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvQyveGAL7af8oA1BT:-KSwiKhL_9_eJ41_0O85:bu3nfni) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.h#L133)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Added citation in C++ implementation.  Should I add it again here?
</details>
It's okay now; marking resolved.

Perhaps I would have just put these in the body of the method that used them (or somehow nearer the citation), but it's reasonably clear for a reader to find as it is now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/examples/Pendulum/pendulum_system.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjRK93zqmt7nk5OXJ:-KSwjRK93zqmt7nk5OXK:b-xcmcas) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L35)):*
> ```objc
>   PendulumStateVector<T>* DoClone() const override;
> 
>   explicit PendulumStateVector(const systems::BasicVector<T>& other) = delete;
> ```

META (next 4 lines) have the wrong argument type.

---

*[drake/examples/Pendulum/pendulum_system.h, line 38 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjH22PW73K69E1XxN:-KSwjH23nFGl2FotkcmI:btd3hjj) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L38)):*
> ```objc
>   PendulumStateVector& operator=(const systems::BasicVector<T>& other) = delete;
>   explicit PendulumStateVector(systems::BasicVector<T>&& other) = delete;
>   PendulumStateVector& operator=(systems::BasicVector<T>&& other) = delete;
> ```

META (entire PR) I am sorry to say, but cppguide says these `= delete` lines must appear in public scope (there's some C++ subtlety involved).

---

*[drake/examples/Pendulum/pendulum_system.h, line 116 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjdHTffx7e6bNdVDi:-KSwjdHTffx7e6bNdVDj:bmo8em3) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L116)):*
> ```objc
>   }
> 
>   explicit PendulumSystem(const systems::LeafSystem<T>& other) = delete;
> ```

META (next 4 lines) incorrect argument type.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 88 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjyxcRqrUSXNB7PLz:-KSwjyxd2NSubV_q4ZeY:b-cw9nmt) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/systems/plants/constraint/direct_collocation_constraint.h#L88)):*
> ```objc
> 
>   explicit SystemDirectCollocationConstraint(
>       const DirectCollocationConstraint& other) = delete;
> ```

META (next 4 lines) Incorrect argument type.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 118 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwk1Nkeg8v3jLDXkWF:-KSwk1Nkeg8v3jLDXkWG:b-cw9nmt) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/systems/plants/constraint/direct_collocation_constraint.h#L118)):*
> ```objc
> 
>   explicit System2DirectCollocationConstraint(
>       const DirectCollocationConstraint& other) = delete;
> ```

META (next 4 lines) Incorrect argument type.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/examples/Pendulum/pendulum_system.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjRK93zqmt7nk5OXJ:-KSwnLoeTujyvOGgHhpR:b-5fm10q) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> META (next 4 lines) have the wrong argument type.
</details>
will fix, but the META isn't confined to here, I got the idea from leaf_system.h

---

*[drake/examples/Pendulum/pendulum_system.h, line 38 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjH22PW73K69E1XxN:-KSwnNBTsqhZWJ0w9_mr:b87x9b1) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L38)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> META (entire PR) I am sorry to say, but cppguide says these `= delete` lines must appear in public scope (there's some C++ subtlety involved).
</details>
will fix, but this is common throughout System2 (see diagram.h, leaf_system.h)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/examples/Pendulum/pendulum_run_dynamics.cc, line 35 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvFf2_snpqeZi148Mc:-KSwv3bqqnKBJvgROZTu:bb77ufg) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_run_dynamics.cc#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Oops, meant to tag @david-german-tri here, but typo'd it.
</details>
Yes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/Pendulum/pendulum_system.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjRK93zqmt7nk5OXJ:-KSxQ19-zWlgpgl7U9Ib:b-peja45) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L35)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> will fix, but the META isn't confined to here, I got the idea from leaf_system.h
</details>
:headdesk:  Probably about 3 months past time that we had a `drake::noncopyable` base class.  I will take that over-arching action item.

---

*[drake/examples/Pendulum/pendulum_system.h, line 38 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjH22PW73K69E1XxN:-KSxQJISAye8brRby2Iz:b-2xy518) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L38)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> will fix, but this is common throughout System2 (see diagram.h, leaf_system.h)
</details>
Yeah; sorry we didn't get it right earlier, we only found that later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Finished first pass from feature review. Some additional documentation would be nice..since this would be the first system2.0 system in Examples. PTAL.

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/examples/Pendulum/CMakeLists.txt, line 14 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KT8ePluragiZi0ezM3U:-KT8ePluragiZi0ezM3V:bx5wi5e) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/CMakeLists.txt#L14)):*
> ```CMake
>     VERSION 0.0.1)
> 
>   add_library_with_exports(LIB_NAME drakePendulumSystem
> ```

BTW: Is it now the fixed convention that all Drake examples also use the drakeFooSystem library name? I was under the impression that is only for DrakeSystem (framework / control / plant) components..

---

*[drake/examples/Pendulum/pendulum_system.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvjgypOTayW6okRret:-KSvjgyr5dBoYtLnqDYf:bml3n3t) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.h#L38)):*
> ```objc
> 
> /// A model of a one-dimensional pendulum systemm.
> ///
> ```

A more detailed description of the Pendulum system (inputs and outputs) might definitely be more useful. Eg. see GravityCompensator

---

*[drake/examples/Pendulum/test/CMakeLists.txt, line 2 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KT8gTtMoDNCMbBOC5Fq:-KT8gTtMoDNCMbBOC5Fr:b7f0yot) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/test/CMakeLists.txt#L2)):*
> ```CMake
> if(lcm_FOUND)
>   add_executable(pendulumURDFDynamicsTest urdfDynamicsTest.cpp)
> ```

Where is this test? 

---

*[drake/examples/Pendulum/test/pendulum_dynamic_constraint_test.cc, line 23 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KT8_zMJfdxW8jpkXLQ8:-KT8_zMJfdxW8jpkXLQ9:b-qaldhs) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/test/pendulum_dynamic_constraint_test.cc#L23)):*
> ```cpp
>   x(3) = 0.05 * M_PI;  // x1(0)
>   x(4) = 0;            // x1(1)
>   x(5) = 0.00537668;   // u0
> ```

BTW: would be nice to have the units here too (just for easier interpretation of the magnitudes)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KT8fkSOOuNTle-NVbSE:bynrcua)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/examples/Pendulum/pendulum_system.h, line 38 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjH22PW73K69E1XxN:-KT9vXA6RGYZkevc3eGO:b-mm15gd) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L38)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Yeah; sorry we didn't get it right earlier, we only found that later.
</details>
Yeah, I've been fixing these as I see them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 6 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/examples/Pendulum/CMakeLists.txt, line 14 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KT8ePluragiZi0ezM3U:-KTAWDX09mtBGS1Gz-KC:bxnqvtl) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/CMakeLists.txt#L14)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> BTW: Is it now the fixed convention that all Drake examples also use the drakeFooSystem library name? I was under the impression that is only for DrakeSystem (framework / control / plant) components..
</details>
I don't think System is necessary for all examples?  I just used the same name as the main exported class.

---

*[drake/examples/Pendulum/pendulum_system.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSvjgypOTayW6okRret:-KTAXW0F4ReuyJ8oLQ7q:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c0afde73ba26643a8f4cf0e64886c895688e9bc/drake/examples/Pendulum/pendulum_system.h#L38)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> A more detailed description of the Pendulum system (inputs and outputs) might definitely be more useful. Eg. see GravityCompensator
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjRK93zqmt7nk5OXJ:-KTAWynGhV9JEFyn31Ny:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> :headdesk:  Probably about 3 months past time that we had a `drake::noncopyable` base class.  I will take that over-arching action item.
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.h, line 38 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjH22PW73K69E1XxN:-KTAX8mJ8FHeDrXDxtMK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L38)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Yeah, I've been fixing these as I see them.
</details>
Done.

---

*[drake/examples/Pendulum/pendulum_system.h, line 116 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjdHTffx7e6bNdVDi:-KTAXXJjK5Fx2UiiyJ25:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/pendulum_system.h#L116)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> META (next 4 lines) incorrect argument type.
</details>
Done.

---

*[drake/examples/Pendulum/test/CMakeLists.txt, line 2 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KT8gTtMoDNCMbBOC5Fq:-KTAXesC6JcVrKrIqTQU:bx6ngnn) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/test/CMakeLists.txt#L2)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Where is this test? 
</details>
I see the referenced .cpp file in the same directory as the CMakeLists.txt, and the output in the expected place in build/drake/bin/pendulumURDFDynamicsTest

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 88 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwjyxcRqrUSXNB7PLz:-KTAYA9L6yVEhbS4CcRM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/systems/plants/constraint/direct_collocation_constraint.h#L88)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> META (next 4 lines) Incorrect argument type.
</details>
Done.

---

*[drake/systems/plants/constraint/direct_collocation_constraint.h, line 118 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KSwk1Nkeg8v3jLDXkWF:-KTAYAk6dYjOa0uEKojp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/systems/plants/constraint/direct_collocation_constraint.h#L118)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> META (next 4 lines) Incorrect argument type.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/Pendulum/test/CMakeLists.txt, line 2 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-KT8gTtMoDNCMbBOC5Fq:-KTAaq3-mpd6LG82puZh:b-fsv86l) ([raw file](https://github.com/robotlocomotion/drake/blob/dd1aa2b459df805e8f7739cac72195428bb11074/drake/examples/Pendulum/test/CMakeLists.txt#L2)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> I see the referenced .cpp file in the same directory as the CMakeLists.txt, and the output in the expected place in build/drake/bin/pendulumURDFDynamicsTest
</details>
Ah so its not part of this PR..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KTBi0ArHchVy3MCz--w:bg7igkv)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KTBkGEpY0pyRb3dDJhi:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
-@amcastro-tri so we don't need a LGTM.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KTBkNZ5I7hEXfkNyyr9:bmagnl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Dismissed Bill, the blocking note was not a defect.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3631#-:-KTBkVoUas7wlEkPfzMJ:b-me60tr)*
<!-- Sent from Reviewable.io -->
